### PR TITLE
gh-115778: Add a ``tierN`` annotation for instruction definition

### PR DESCRIPTION
**Pull request by Eclips4**
* Issue: gh-115778

### CODE REVIEW

File: Python/bytecodes.c [hunk 1]
```diff
@@ -318,8 +315,7 @@ dummy_func(
             TO_BOOL_STR,
         };
 
-        specializing op(_SPECIALIZE_TO_BOOL, (counter/1, value -- value)) {
-            TIER_ONE_ONLY
+        tier1 specializing op(_SPECIALIZE_TO_BOOL, (counter/1, value -- value)) {
```
Comment (markshannon) [inline Python/bytecodes.c]: `specializing` implies that it is tier 1 only, so maybe drop the `tier1` here, and have the code generator handle it?

File: Tools/cases_generator/lexer.py [hunk 2]
```diff
@@ -288,7 +288,7 @@ def tokenize(src: str, line: int = 1, filename: str = "") -> Iterator[Token]:
         text = m.group(0)
         if text in keywords:
             kind = keywords[text]
-        elif text in annotations:
+        elif text in annotations or "tier" in text:
```
Comment (markshannon) [inline Tools/cases_generator/lexer.py]: Add "tier1" and "tier2" to annotations, otherwise this will match "fattier"

**Review by markshannon — COMMENTED** [review]
Looks like a nice improvement. Thanks.

File: Python/bytecodes.c (hunk 1 above)
Comment (Eclips4) [reply Python/bytecodes.c]: Sounds reasonable. Done

File: Tools/cases_generator/tier2_abstract_generator.py [hunk 3]
```diff
@@ -176,7 +176,7 @@ def generate_abstract_interpreter(
         if uop.name in abstract.uops:
             override = abstract.uops[uop.name]
             validate_uop(override, uop)
-        if uop.properties.tier_one_only:
+        if uop.properties.tier == 1 or "specializing" in uop.annotations:
```
Comment (gvanrossum) [inline Tools/cases_generator/tier2_abstract_generator.py]: Hm, given that this change occurs in three different files, I would have solved this differently -- when the uop is created, if `specializing` is given, force `tier = 1`. Unless this affects metadata generation?

File: Tools/cases_generator/analyzer.py [hunk 4]
```diff
@@ -312,6 +310,12 @@ def variable_used(node: parser.InstDef, name: str) -> bool:
         token.kind == "IDENTIFIER" and token.text == name for token in node.tokens
     )
 
+def tier_variable(node: parser.InstDef) -> int | None:
+    """Determine whether a tier variable is used in a node."""
+    for token in node.tokens:
+        if token.kind == "ANNOTATION" and re.match(r"tier\d", token.text):
+            return int(token.text[-1])
```
Comment (gvanrossum) [inline Tools/cases_generator/analyzer.py]: (This is where you could force `tier1` if `specialized` is set.)

**Review by gvanrossum — COMMENTED** [review]
Basically LG, though I have two nits:
- If there ever is a Tier 3, I expect that several of the tier checks will have to be adjusted -- currently there seems to be an assumption that if it's not T1 then it's T2 and vice versa.
- Inferring tier 1 from specialized could be done differently.

File: Tools/cases_generator/analyzer.py [hunk 5]
```diff
@@ -312,6 +310,15 @@ def variable_used(node: parser.InstDef, name: str) -> bool:
         token.kind == "IDENTIFIER" and token.text == name for token in node.tokens
     )
 
+def tier_variable(node: parser.InstDef) -> int | None:
+    """Determine whether a tier variable is used in a node."""
+    for token in node.tokens:
+        if token.kind == "ANNOTATION":
+            if token.text == "specializing":
+                return 1
+            if re.match(r"tier\d", token.text):
```
Comment (AlexWaygood) [inline Tools/cases_generator/analyzer.py]:
```suggestion
            if re.fullmatch(r"tier\d", token.text):
```

**Review by gvanrossum — APPROVED** [review]
Thanks!

**Comment by Eclips4** [comment]
@gvanrossum 
CI/CD failure is unrelated. I can reproduce it locally on my Windows x86_64 build with JIT. I'll create a issue

**Comment by gvanrossum** [comment]
Thanks!

**Pull request merged**